pepper-apis/src/main/java/org/broadinstitute/ddp/jetty/JettyConfig.java

```diff
@@ -0,0 +1,153 @@
+            this.threadPool = threadPool;
+            return this;
+        }
+
+        public AppEngineServerFactory withHttpOnly(boolean httpOnly) {
+            this.httpOnly = httpOnly;
+            return this;
+        }
+    }
+
+    /**
+     * Lifted from {@link spark.embeddedserver.jetty.JettyHandler}, with one
+     * change to use {@link SourceIPHeaderRequestWrapper} to set
+     * the source IP properly in the appengine environment
+     */
+    public static class JettyCustomRemoteAddrHeaderHandler extends SessionHandler {
+        private javax.servlet.Filter filter;
+        private String preferredIPSourceHeader;
+
+        public JettyCustomRemoteAddrHeaderHandler(Filter filter, String preferredIPSourceHeader) {
+            this.filter = filter;
+            this.preferredIPSourceHeader = preferredIPSourceHeader;
+        }
+
+        public void doHandle(String target, org.eclipse.jetty.server.Request baseRequest, HttpServletRequest request,
+                             HttpServletResponse response) throws IOException, ServletException {
+            HttpRequestWrapper wrapper = new SourceIPHeaderRequestWrapper(request, preferredIPSourceHeader);
+            this.filter.doFilter(wrapper, response, null);
+            if (wrapper.notConsumed()) {
+                baseRequest.setHandled(false);
+            } else {
+                baseRequest.setHandled(true);
+            }
+        }
+    }
+
+    /**
+     * Lifted from {@link spark.embeddedserver.jetty.JettyServer}
+     */
+    public static class JettyServer implements JettyServerFactory {
+        JettyServer() {
+        }
+
+        public Server create(int maxThreads, int minThreads, int threadTimeoutMillis) {
+            Server server;
+            if (maxThreads > 0) {
+                int min = minThreads > 0 ? minThreads : 8;
+
+                int idleTimeout = threadTimeoutMillis > 0 ? threadTimeoutMillis : 30_000;
+                server = new Server(new QueuedThreadPool(maxThreads, min, idleTimeout));
+            } else {
+                server = new Server();
+            }
+
+            return server;
+        }
+
+        public Server create(ThreadPool threadPool) {
+            return threadPool != null ? new Server(threadPool) : new Server();
+        }
+    }
+
+    /**
+     * RequestWrapper that uses a different header as the preferred source
+     * for the ip address that is returned by {@link HttpServletRequest#getRemoteAddr()}.
+     */
+    public static class SourceIPHeaderRequestWrapper extends HttpRequestWrapper {
+
+        String remoteAddr;
+
+        public SourceIPHeaderRequestWrapper(HttpServletRequest request, String preferredHeaderForIP) {
+            super(request);
+            if (StringUtils.isNotBlank(preferredHeaderForIP)) {
+                String appEngineSourceIp = request.getHeader(preferredHeaderForIP);
+                remoteAddr = request.getRemoteAddr();
```

Comment:
I think we need to move this line to above the `if` block. I'm trying to run things locally, and when I don't have the `preferredHeaderForIP`, `remoteAddr` is not being set and I'm hitting a NPE.